File: lc_soln/0036_valid_sudoku/solution.py
from typing import List

SIZE = 9
BSIZE = 3


def blk_idx(row: int, col: int) -> int:
    return (row // BSIZE) * BSIZE + (col // BSIZE)


class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        rows = [0] * SIZE
        cols = [0] * SIZE
        blks = [0] * SIZE
        for i in range(SIZE):
            for j in range(SIZE):
                if board[i][j] == ".":
                    continue
                offset = 1 << int(board[i][j])
                if (
                    (rows[i] & offset)
                    or (cols[j] & offset)
                    or (blks[blk_idx(i, j)] & offset)
                ):
                    return False
                rows[i] |= offset
                cols[j] |= offset
                blks[blk_idx(i, j)] |= offset

        return True


def main(): ...


if __name__ == "__main__":
    main()

File: lc_soln/0036_valid_sudoku/solution.cpp
const int SIZE = 9;
const int BSIZE = 3;

int blk_idx(int i, int j) { return (i / BSIZE) * BSIZE + j / BSIZE; }

class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        std::vector<int> rows(SIZE, 0);
        std::vector<int> cols(SIZE, 0);
        std::vector<int> blks(SIZE, 0);
        for (int i = 0; i < SIZE; ++i) {
            for (int j = 0; j < SIZE; ++j) {
                if (board[i][j] == '.') {
                    continue;
                }
                int offset = 1 << (board[i][j] - '0');
                if ((rows[i] & offset) != 0 || (cols[j] & offset) != 0 ||
                    (blks[blk_idx(i, j)] & offset) != 0) {
                    return false;
                }
                rows[i] |= offset;
                cols[j] |= offset;
                blks[blk_idx(i, j)] |= offset;
            }
        }
        return true;
    }
};

int main() {
    return 0;
}
